File: backend/wire-api/src/services/riskScoring.ts
/**
 * Risk Scoring Service
 * 
 * Calculates risk scores for intents
 * Integrates with Fraud Detection Service
 */

import { getDatabase } from '../db/connection';
import { logger } from '../utils/logger';

const FRAUD_SERVICE_URL = process.env.FRAUD_SERVICE_URL || 'http://localhost:8002';

export interface RiskScore {
  score: number;
  rationale: {
    factors: Array<{ name: string; score: number; details: any }>;
    scoreBreakdown: {
      base: number;
      amount?: number;
      beneficiary?: number;
      rails?: number;
      international?: number;
      timing?: number;
      relationship?: number;
      anomalies?: number;
      patterns?: number;
      total: number;
    };
  };
}

export interface RiskScoringInput {
  amountMinor: number;
  currency: string;
  railsType: string;
  beneficiaryId: string;
  createdAt: Date;
  userId: string;
}

/**
 * Calculate risk score for an intent
 */
export async function calculateRiskScore(
  input: RiskScoringInput,
  beneficiary: any,
  user: any
): Promise<RiskScore> {
  const factors: Array<{ name: string; score: number; details: any }> = [];
  let totalScore = 0;

  // Amount risk
  const amountRisk = calculateAmountRisk(input.amountMinor);
  totalScore += amountRisk.score;
  factors.push({ name: 'amount', score: amountRisk.score, details: amountRisk });

  // Beneficiary risk
  const beneficiaryRisk = await calculateBeneficiaryRisk(beneficiary);
  totalScore += beneficiaryRisk.score;
  factors.push({ name: 'beneficiary', score: beneficiaryRisk.score, details: beneficiaryRisk });

  // User risk
  const userRisk = await calculateUserRisk(user);
  totalScore += userRisk.score;
  factors.push({ name: 'user', score: userRisk.score, details: userRisk });

  // Rails risk
  const railsRisk = calculateRailsRisk(input.railsType);
  totalScore += railsRisk.score;
  factors.push({ name: 'rails', score: railsRisk.score, details: railsRisk });

  // International risk
  const internationalRisk = calculateInternationalRisk(beneficiary);
  totalScore += internationalRisk.score;
  factors.push({ name: 'international', score: internationalRisk.score, details: internationalRisk });

  // Timing risk
  const timingRisk = calculateTimingRisk(input.createdAt);
  totalScore += timingRisk.score;
  factors.push({ name: 'timing', score: timingRisk.score, details: timingRisk });

  // Relationship risk
  const relationshipRisk = await calculateRelationshipRisk(input.userId, input.beneficiaryId);
  totalScore += relationshipRisk.score;
  factors.push({ name: 'relationship', score: relationshipRisk.score, details: relationshipRisk });

  // Anomaly detection
  const anomalies = await detectAnomalies(input, beneficiary, user);
  const anomalyScore = anomalies.reduce((sum, a) => sum + a.severity, 0);
  totalScore += anomalyScore;
  factors.push({ name: 'anomalies', score: anomalyScore, details: anomalies });

  // Pattern analysis
  const patterns = await analyzePatterns(input.userId);
  totalScore += patterns.riskScore;
  factors.push({ name: 'patterns', score: patterns.riskScore, details: patterns });

  // Cap at 100
  const finalScore = Math.min(100, Math.max(0, totalScore));

  return {
    score: finalScore,
    rationale: {
      factors,
      scoreBreakdown: {
        base: 0,
        amount: amountRisk.score,
        beneficiary: beneficiaryRisk.score,
        rails: railsRisk.score,
        international: internationalRisk.score,
        timing: timingRisk.score,
        relationship: relationshipRisk.score,
        anomalies: anomalyScore,
        patterns: patterns.riskScore,
        total: finalScore,
      },
    },
  };
}

function calculateAmountRisk(amountMinor: number): { score: number; details: any } {
  const amountDollars = amountMinor / 100;
  let score = 0;

  if (amountDollars >= 1000000) {
    score = 30; // Very high amount
  } else if (amountDollars >= 500000) {
    score = 20; // High amount
  } else if (amountDollars >= 100000) {
    score = 10; // Moderate amount
  } else if (amountDollars >= 50000) {
    score = 5; // Low-moderate amount
  }

  return {
    score,
    details: {
      amountDollars,
      threshold: amountDollars >= 100000 ? 'high' : amountDollars >= 50000 ? 'moderate' : 'low',
    },
  };
}

async function calculateBeneficiaryRisk(beneficiary: any): Promise<{ score: number; details: any }> {
  let score = 0;

  // New beneficiary risk
  const db = getDatabase();
  const beneficiaryAgeResult = await db.query(
    'SELECT EXTRACT(EPOCH FROM (NOW() - created_at)) / 86400 as age_days FROM beneficiaries WHERE id = $1',
    [beneficiary.id]
  );

  const ageDays = beneficiaryAgeResult.rows[0]?.age_days || 0;
  if (ageDays < 30) {
    score += 15; // New beneficiary
  } else if (ageDays < 90) {
    score += 5; // Recent beneficiary
  }

  // Verification status
  if (beneficiary.verification_status !== 'VERIFIED') {
    score += 10;
  }

  // KYC status
  if (beneficiary.kyc_status !== 'PASSED') {
    score += 10;
  }

  // Sanctions check
  if (beneficiary.sanctions_check_status === 'FLAGGED') {
    score += 50; // Critical
  }

  return {
    score,
    details: {
      ageDays,
      verificationStatus: beneficiary.verification_status,
      kycStatus: beneficiary.kyc_status,
      sanctionsStatus: beneficiary.sanctions_check_status,
    },
  };
}

async function calculateUserRisk(user: any): Promise<{ score: number; details: any }> {
  let score = 0;

  // Account age
  const db = getDatabase();
  const userAgeResult = await db.query(
    'SELECT EXTRACT(EPOCH FROM (NOW() - created_at)) / 86400 as age_days FROM users WHERE id = $1',
    [user.id]
  );

  const ageDays = userAgeResult.rows[0]?.age_days || 0;
  if (ageDays < 30) {
    score += 10; // New user
  }

  // Failed login attempts
  if (user.failed_login_attempts > 0) {
    score += user.failed_login_attempts * 2;
  }

  // MFA status
  if (!user.mfa_enabled) {
    score += 5;
  }

  return {
    score,
    details: {
      ageDays,
      failedLoginAttempts: user.failed_login_attempts,
      mfaEnabled: user.mfa_enabled,
    },
  };
}

function calculateRailsRisk(railsType: string): { score: number; details: any } {
  // WIRE transfers are higher risk than ACH
  return {
    score: railsType === 'WIRE' ? 10 : 5,
    details: { railsType },
  };
}

function calculateInternationalRisk(beneficiary: any): { score: number; details: any } {
  // International transfers are higher risk
  const isInternational = beneficiary.country !== 'US';
  return {
    score: isInternational ? 15 : 0,
    details: {
      country: beneficiary.country,
      isInternational,
    },
  };
}

function calculateTimingRisk(createdAt: Date): { score: number; details: any } {
  const hour = createdAt.getHours();
  const dayOfWeek = createdAt.getDay();

  let score = 0;

  // Off-hours risk (outside 9 AM - 5 PM)
  if (hour < 9 || hour >= 17) {
    score += 5;
  }

  // Weekend risk
  if (dayOfWeek === 0 || dayOfWeek === 6) {
    score += 5;
  }

  return {
    score,
    details: {
      hour,
      dayOfWeek,
      isOffHours: hour < 9 || hour >= 17,
      isWeekend: dayOfWeek === 0 || dayOfWeek === 6,
    },
  };
}

async function calculateRelationshipRisk(
  userId: string,
  beneficiaryId: string
): Promise<{ score: number; details: any }> {
  const db = getDatabase();

  // Check history of transfers to this beneficiary
  const historyResult = await db.query(
    `SELECT COUNT(*) as count, SUM(amount_minor) as total_amount
     FROM intents
     WHERE created_by_user_id = $1 AND beneficiary_id = $2
     AND status IN ('EXECUTED', 'APPROVED')
     AND created_at > NOW() - INTERVAL '90 days'`,
    [userId, beneficiaryId]
  );

  const count = parseInt(historyResult.rows[0]?.count || '0');
  const totalAmount = parseInt(historyResult.rows[0]?.total_amount || '0');

  let score = 0;

  // New relationship
  if (count === 0) {
    score += 10;
  }

  return {
    score,
    details: {
      transferCount: count,
      totalAmount,
      isNewRelationship: count === 0,
    },
  };
}

async function detectAnomalies(
  input: RiskScoringInput,
  beneficiary: any,
  user: any
): Promise<Array<{ type: string; severity: number; details: any }>> {
  const anomalies: Array<{ type: string; severity: number; details: any }> = [];
  const db = getDatabase();

  // Check for rapid-fire transfers
  const recentTransfersResult = await db.query(
    `SELECT COUNT(*) as count
     FROM intents
     WHERE created_by_user_id = $1
     AND created_at > NOW() - INTERVAL '1 hour'
     AND status != 'CANCELED'`,
    [input.userId]
  );

  const recentCount = parseInt(recentTransfersResult.rows[0]?.count || '0');
  if (recentCount > 5) {
    anomalies.push({
      type: 'RAPID_FIRE_TRANSFERS',
      severity: 15,
      details: { count: recentCount },
    });
  }

  // Check for unusual amount
  const userHistoryResult = await db.query(
    `SELECT AVG(amount_minor) as avg_amount, MAX(amount_minor) as max_amount
     FROM intents
     WHERE created_by_user_id = $1
     AND status IN ('EXECUTED', 'APPROVED')
     AND created_at > NOW() - INTERVAL '90 days'`,
    [input.userId]
  );

  const avgAmount = parseFloat(userHistoryResult.rows[0]?.avg_amount || '0');
  const maxAmount = parseFloat(userHistoryResult.rows[0]?.max_amount || '0');

  if (avgAmount > 0 && input.amountMinor > avgAmount * 3) {
    anomalies.push({
      type: 'UNUSUAL_AMOUNT',
      severity: 10,
      details: {
        amount: input.amountMinor,
        averageAmount: avgAmount,
        multiplier: input.amountMinor / avgAmount,
      },
    });
  }

  return anomalies;
}

async function analyzePatterns(userId: string): Promise<{ riskScore: number; details: any }> {
  const db = getDatabase();

  // Analyze user's transfer patterns
  const patternResult = await db.query(
    `SELECT 
       COUNT(*) as total_transfers,
       COUNT(DISTINCT beneficiary_id) as unique_beneficiaries,
       SUM(amount_minor) as total_amount,
       AVG(amount_minor) as avg_amount
     FROM intents
     WHERE created_by_user_id = $1
     AND created_at > NOW() - INTERVAL '30 days'
     AND status IN ('EXECUTED', 'APPROVED')`,
    [userId]
  );

  const patterns = patternResult.rows[0] || {};
  let riskScore = 0;

  // High number of unique beneficiaries
  if (parseInt(patterns.unique_beneficiaries || '0') > 20) {
    riskScore += 5;
  }

  // Very high total amount
  const totalAmount = parseFloat(patterns.total_amount || '0');
  if (totalAmount > 10000000) {
    // $100k
    riskScore += 10;
  }

  return {
    riskScore,
    details: patterns,
  };
}

File: backend/fraud-service/services/risk_scoring.py
"""
Risk Scoring Service
Calculates comprehensive risk scores for intents
"""

import json
from typing import Dict, Any, List
from datetime import datetime, timedelta
from services.anomaly_detection import detect_anomalies
from services.pattern_analysis import analyze_patterns
from services.behavioral_analysis import analyze_behavior
from db.connection import get_db_connection
from utils.logger import get_logger

logger = get_logger(__name__)

class RiskScoringService:
    """Risk scoring service"""
    
    def __init__(self):
        self.max_score = 100
    
    async def calculate_risk_score(
        self,
        intent_data: Dict[str, Any],
        beneficiary_data: Dict[str, Any],
        user_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Calculate comprehensive risk score for intent
        
        Args:
            intent_data: Intent data
            beneficiary_data: Beneficiary data
            user_data: User data
            
        Returns:
            Risk score with factors and rationale
        """
        logger.info(f"Calculating risk score for intent {intent_data.get('id')}")
        
        factors = []
        score = 0
        
        # 1. Amount risk
        amount_risk = self._calculate_amount_risk(intent_data)
        score += amount_risk['score']
        factors.append({
            'factor': 'amount',
            'score': amount_risk['score'],
            'details': amount_risk
        })
        
        # 2. Beneficiary risk
        beneficiary_risk = await self._calculate_beneficiary_risk(
            beneficiary_data,
            intent_data
        )
        score += beneficiary_risk['score']
        factors.append({
            'factor': 'beneficiary',
            'score': beneficiary_risk['score'],
            'details': beneficiary_risk
        })
        
        # 3. User risk
        user_risk = await self._calculate_user_risk(user_data, intent_data)
        score += user_risk['score']
        factors.append({
            'factor': 'user',
            'score': user_risk['score'],
            'details': user_risk
        })
        
        # 4. Timing risk
        timing_risk = self._calculate_timing_risk(intent_data)
        score += timing_risk['score']
        factors.append({
            'factor': 'timing',
            'score': timing_risk['score'],
            'details': timing_risk
        })
        
        # 5. Relationship risk
        relationship_risk = await self._calculate_relationship_risk(
            intent_data.get('created_by_user_id'),
            intent_data.get('beneficiary_id'),
            intent_data
        )
        score += relationship_risk['score']
        factors.append({
            'factor': 'relationship',
            'score': relationship_risk['score'],
            'details': relationship_risk
        })
        
        # 6. Anomaly detection
        anomalies = await detect_anomalies(intent_data, beneficiary_data, user_data)
        anomaly_score = sum(a.get('severity', 0) for a in anomalies)
        score += anomaly_score
        factors.append({
            'factor': 'anomalies',
            'score': anomaly_score,
            'count': len(anomalies),
            'details': anomalies
        })
        
        # 7. Pattern analysis
        patterns = await analyze_patterns(
            intent_data.get('created_by_user_id'),
            intent_data
        )
        score += patterns.get('risk_score', 0)
        factors.append({
            'factor': 'patterns',
            'score': patterns.get('risk_score', 0),
            'details': patterns
        })
        
        # 8. Behavioral analysis
        behavior = await analyze_behavior(
            intent_data.get('created_by_user_id'),
            intent_data
        )
        score += behavior.get('risk_score', 0)
        factors.append({
            'factor': 'behavior',
            'score': behavior.get('risk_score', 0),
            'details': behavior
        })
        
        # Cap score at maximum
        final_score = min(self.max_score, score)
        
        # Generate rationale
        rationale = self._generate_rationale(factors, final_score)
        
        # Generate recommendations
        recommendations = self._generate_recommendations(final_score, factors)
        
        logger.info(
            f"Risk score calculated: {final_score}/100 for intent {intent_data.get('id')}"
        )
        
        return {
            'score': final_score,
            'factors': factors,
            'rationale': rationale,
            'recommendations': recommendations,
            'calculated_at': datetime.utcnow().isoformat()
        }
    
    def _calculate_amount_risk(self, intent_data: Dict[str, Any]) -> Dict[str, Any]:
        """Calculate amount-based risk"""
        amount_minor = intent_data.get('amount_minor', 0)
        amount_dollars = amount_minor / 100
        
        # Risk tiers
        if amount_dollars >= 1000000:  # $1M+
            risk_score = 30
            risk_level = 'VERY_HIGH'
        elif amount_dollars >= 500000:  # $500K+
            risk_score = 20
            risk_level = 'HIGH'
        elif amount_dollars >= 100000:  # $100K+
            risk_score = 10
            risk_level = 'MEDIUM'
        elif amount_dollars >= 50000:  # $50K+
            risk_score = 5
            risk_level = 'LOW'
        else:
            risk_score = 0
            risk_level = 'VERY_LOW'
        
        return {
            'score': risk_score,
            'amount_dollars': amount_dollars,
            'risk_level': risk_level
        }
    
    async def _calculate_beneficiary_risk(
        self,
        beneficiary_data: Dict[str, Any],
        intent_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Calculate beneficiary-based risk"""
        if not beneficiary_data:
            return {'score': 0, 'reason': 'No beneficiary data'}
        
        risk_score = 0
        risk_factors = []
        
        # New beneficiary (not verified)
        if beneficiary_data.get('verification_status') != 'VERIFIED':
            risk_score += 20
            risk_factors.append('unverified_beneficiary')
        
        # New beneficiary (recently created)
        created_at = beneficiary_data.get('created_at')
        if created_at:
            created_date = datetime.fromisoformat(created_at.replace('Z', '+00:00'))
            days_old = (datetime.utcnow() - created_date.replace(tzinfo=None)).days
            if days_old < 7:
                risk_score += 15
                risk_factors.append('new_beneficiary')
        
        # Locked beneficiary
        if beneficiary_data.get('status') == 'LOCKED':
            risk_score += 50
            risk_factors.append('locked_beneficiary')
        
        # Sanctions check
        if beneficiary_data.get('sanctions_check_status') == 'FLAGGED':
            risk_score += 85
            risk_factors.append('sanctions_flagged')
        
        # No transfer history
        db = await get_db_connection()
        transfer_count = await db.fetchval(
            """
            SELECT COUNT(*) FROM intents
            WHERE beneficiary_id = $1 AND status = 'EXECUTED'
            """,
            [beneficiary_data.get('id')]
        )
        
        if transfer_count == 0:
            risk_score += 10
            risk_factors.append('no_transfer_history')
        
        return {
            'score': min(85, risk_score),  # Cap beneficiary risk
            'risk_factors': risk_factors,
            'beneficiary_id': beneficiary_data.get('id')
        }
    
    async def _calculate_user_risk(
        self,
        user_data: Dict[str, Any],
        intent_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Calculate user-based risk"""
        if not user_data:
            return {'score': 0, 'reason': 'No user data'}
        
        risk_score = 0
        risk_factors = []
        
        # New user
        created_at = user_data.get('created_at')
        if created_at:
            created_date = datetime.fromisoformat(created_at.replace('Z', '+00:00'))
            days_old = (datetime.utcnow() - created_date.replace(tzinfo=None)).days
            if days_old < 30:
                risk_score += 10
                risk_factors.append('new_user')
        
        # User role
        role = user_data.get('role', '')
        if role not in ['TREASURY_INITIATOR', 'ADMIN']:
            risk_score += 5
            risk_factors.append('non_standard_role')
        
        # Recent failed attempts
        db = await get_db_connection()
        recent_failures = await db.fetchval(
            """
            SELECT COUNT(*) FROM intents
            WHERE created_by_user_id = $1
            AND status = 'DENIED'
            AND created_at > NOW() - INTERVAL '7 days'
            """,
            [user_data.get('id')]
        )
        
        if recent_failures > 3:
            risk_score += 15
            risk_factors.append('recent_failures')
        
        return {
            'score': min(20, risk_score),  # Cap user risk
            'risk_factors': risk_factors,
            'user_id': user_data.get('id')
        }
    
    def _calculate_timing_risk(self, intent_data: Dict[str, Any]) -> Dict[str, Any]:
        """Calculate timing-based risk"""
        created_at = intent_data.get('created_at')
        if not created_at:
            return {'score': 0}
        
        created_date = datetime.fromisoformat(created_at.replace('Z', '+00:00'))
        hour = created_date.hour
        day_of_week = created_date.weekday()
        
        risk_score = 0
        risk_factors = []
        
        # Off-hours (outside 9 AM - 5 PM)
        if hour < 9 or hour >= 17:
            risk_score += 5
            risk_factors.append('off_hours')
        
        # Weekend
        if day_of_week >= 5:  # Saturday or Sunday
            risk_score += 5
            risk_factors.append('weekend')
        
        return {
            'score': min(10, risk_score),  # Cap timing risk
            'risk_factors': risk_factors,
            'hour': hour,
            'day_of_week': day_of_week
        }
    
    async def _calculate_relationship_risk(
        self,
        user_id: str,
        beneficiary_id: str,
        intent_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Calculate relationship-based risk"""
        if not user_id or not beneficiary_id:
            return {'score': 0}
        
        db = await get_db_connection()
        
        # Check transfer history between user and beneficiary
        transfer_count = await db.fetchval(
            """
            SELECT COUNT(*) FROM intents
            WHERE created_by_user_id = $1
            AND beneficiary_id = $2
            AND status = 'EXECUTED'
            """,
            [user_id, beneficiary_id]
        )
        
        risk_score = 0
        risk_factors = []
        
        # No previous transfers
        if transfer_count == 0:
            risk_score += 10
            risk_factors.append('no_previous_transfers')
        
        # First transfer in 30 days
        last_transfer = await db.fetchval(
            """
            SELECT MAX(executed_at) FROM intents
            WHERE created_by_user_id = $1
            AND beneficiary_id = $2
            AND status = 'EXECUTED'
            """,
            [user_id, beneficiary_id]
        )
        
        if last_transfer:
            last_date = datetime.fromisoformat(str(last_transfer).replace('Z', '+00:00'))
            days_since = (datetime.utcnow() - last_date.replace(tzinfo=None)).days
            if days_since > 30:
                risk_score += 5
                risk_factors.append('stale_relationship')
        
        return {
            'score': min(10, risk_score),  # Cap relationship risk
            'risk_factors': risk_factors,
            'transfer_count': transfer_count
        }
    
    def _generate_rationale(
        self,
        factors: List[Dict[str, Any]],
        final_score: int
    ) -> str:
        """Generate human-readable rationale"""
        high_risk_factors = [
            f for f in factors
            if f.get('score', 0) >= 10
        ]
        
        if final_score >= 80:
            level = "CRITICAL"
        elif final_score >= 60:
            level = "HIGH"
        elif final_score >= 40:
            level = "MEDIUM"
        elif final_score >= 20:
            level = "LOW"
        else:
            level = "VERY_LOW"
        
        rationale = f"Risk score: {final_score}/100 ({level}). "
        
        if high_risk_factors:
            top_factors = sorted(
                high_risk_factors,
                key=lambda x: x.get('score', 0),
                reverse=True
            )[:3]
            factor_names = [f['factor'] for f in top_factors]
            rationale += f"Primary risk factors: {', '.join(factor_names)}."
        
        return rationale
    
    def _generate_recommendations(
        self,
        score: int,
        factors: List[Dict[str, Any]]
    ) -> List[str]:
        """Generate recommendations based on risk score"""
        recommendations = []
        
        if score >= 80:
            recommendations.append("CRITICAL: Require additional approvals")
            recommendations.append("CRITICAL: Manual review required")
            recommendations.append("CRITICAL: Consider blocking transfer")
        elif score >= 60:
            recommendations.append("HIGH: Require additional approvals")
            recommendations.append("HIGH: Enhanced verification recommended")
        elif score >= 40:
            recommendations.append("MEDIUM: Standard approval process")
        elif score >= 20:
            recommendations.append("LOW: Standard approval process")
        else:
            recommendations.append("VERY_LOW: Standard approval process")
        
        # Factor-specific recommendations
        for factor in factors:
            factor_name = factor.get('factor')
            factor_score = factor.get('score', 0)
            
            if factor_name == 'beneficiary' and factor_score >= 20:
                recommendations.append("Verify beneficiary identity")
            
            if factor_name == 'anomalies' and factor.get('count', 0) > 0:
                recommendations.append("Review detected anomalies")
            
            if factor_name == 'amount' and factor_score >= 20:
                recommendations.append("Verify large amount transfer")
        
        return recommendations

risk_scoring_service = RiskScoringService()
